feat: Changed typings from unknown to something a little more useful

alexasselin008 opened the pull request:
Advantages: 
- Ensures that the runtime passed to the register method extends from AbstractRuntime
-`registerStaticModules` now makes sure the register method has the same types as the runtime and context passed with it
- Added typings for the `Container`, `Factory` and `Module` in order to prevent handling any everywhere in the loadRemove code

patricklafrance commented:
I might have missed a few changes, it's hard to tell because it seems intertwined with a merge conflicts.

However, the types now seems more complicated and I can't say that I see much benefit to it. I can go more in details about this if you want to, given that you clean up the PR to remove the changes unrelated to this PR but in short, most of the changes affects internal code for which improving the types have little to no benefit IMO.

alexasselin008 commented:
@patricklafrance i updated the PR and fixed the conflicts in order to simplify the review

I added some typings "internally" to the loadRemote method. Previously, it was returning "any" and that could lead to potential mistakes made by maintainers when it's used in other files. So, by adding typings, we can make sure we avoid those mistakes and improve type safety.

I also made some changes to the typings for the register method. Now, it validates that the runtime passed in is extending the AbstractRuntime type. This means that we don't have to rely on the comment you added in the typings anymore, which mentioned using "any" instead of "unknown" for casting. We don't want to deal with implicit casting in parameter declarations.

patricklafrance commented on packages/webpack-module-federation/src/loadRemote.ts:
```diff
@@ -1,7 +1,17 @@
+
+// The interface of a module that can be loaded dynamically.
+interface Module {
+    register?: ModuleRegisterFunction;
```
I am not sure about this. At some point something else could be loaded than our standard register function. It's up to the consumer.

alexasselin008 replied:
The idea is we own the code that allows us to register remote modules into the host. in the `registerRemoteModules` we already look for `module.register`. The client can register whatever they want, be we still need to manipulate those registration here, so it make sense to define what we are expecting. 

the `registerRemoteModules` file isnt shown in this review because it hasn't changed, but it uses what has been registered, so i really think this type is necessary

patricklafrance commented on packages/webpack-module-federation/src/loadRemote.ts:
```diff
@@ -1,7 +1,17 @@
+interface Container {
+    init: (shareScope: unknown) => void;
+    get: (module: string) => () => Module;
+}
```
I don't think I would add this. This code is mostly copied from Webpack docs and this is not a piece of code that I expect us to evolve on our own. I rather copy/paste updated versions from the Webpack docs.

alexasselin008 replied:
I think its fair. i only kept the method's return value instead of the whole container
